### ISSUE TITLE
Bug: [no-unused-vars] should not warn for variables of type 'never'

### ISSUE THREAD

**Issue by dmurvihill** [typescript-eslint/typescript-eslint#11781]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=FASwdgLgpgTgZgQwMZQAQBUCeAHKBGVAb2FVQhygC5UByc3AWjxoG4TUBnAewFsoARBBATUOEGOADmbAL7BQkWIhQYKAJiLt6VWtoZrW7LhAAWsQcNHips%2BdtQBVMCC5gsuVAF5VuAgB8fKDV5OABXMCQIFzBUABMuAGVeKCdo9ygEiFC4OAAKLmpU13SASk1SEDhUfIA6e08G3QomGjLiUlIAek64rjIKAjFsuHYZVCgAGw40SuquOoovRrpmgzb2Lp74-twNIZzR8am0do7u1EQQCZ20JDMkAGspVFmAd1uuUInY8YA3WFQACM0GYYFANqgkK4xKgAPrUMBQf4wLyoLhsUhyGRAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
interface Type1 {
  type: 'type-1';
  someData: string;
}

interface Type2 {
  type: 'type-2';
  otherData: string;
}

type UnionType = Type1 | Type2

function doSomeUnionTypeStuff(o: UnionType) {
  if (o.type === 'type-1') {
    // do type1 stuff
  } else if (o.type === 'type-2') {
    // do type2 stuff
  } else {
    // fail type checking if we could ever be here
    const _: never = o;
  }
}
```

### ESLint Config

```javascript
{
  "rules": {}
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

```json
No lint warnings
```

### Actual Result

```
6133: '_' is declared but its value is never read. 21:11 - 21:12
```

### Additional Info

The unused variable is a bit hacky, but it's an elegant way to get compile-time behavior similar to the `match` feature in Rust and other languages. Very nice for managing union types, especially errors when working in an EAV paradigm. When using this construct, you can be sure that adding more cases to your union type won't cause switches to miss anywhere it's used.

Conversely, it's hard for me to imagine a case where someone would accidentally define an unused variable of type `never`. This is a pattern that's used mainly by people who know what they are doing and why.

Please consider carving variables of type `never` out of `no-unused-vars`.

**Comment by bradzacher** [comment]
I don't think this is something we want to provide:
- Unused variables are generally an anti-pattern so adding "first class support" for always ignoring provably unused variables doesn't seem like it aligns with the rule
- There is already a mechanism through which you can ignore variables -- using the ignore pattern settings. From my experience many codebases use this option to ignore variables prefixed with `_`.
- A lot of codebases implement an "unreachable error" with the signature `constructor(value: never)` so they can do `throw new UnreachableError(o)` that not only provides the same typecheck time validation -- but also provides runtime validation to prevent "undefined behaviour".
- And most importantly -- there is a better way to achieve this same functionality without any variable at all -- [`o satisfies never`](https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=FASwdgLgpgTgZgQwMZQAQBUCeAHKBGVAb2FVQhygC5UByc3AWjxoG4TUBnAewFsoARBBATUOEGOADmbAL7BQkWIhQYKAJiLt6VWtoZrW7LhAAWsQcNHips%2BdtQBVMCC5gsuVAF5VuAgB8fKDV5OABXMCQIFzBUABMuAGVeKCdo9ygEiFC4OAAKLmpU13SASk1SEDhUfIA6e08G3QomGjLiUlIAek64rjIKAjFsuHZu1BlUKAAbDjRK6q46ii9GumaDNvYunvj%2B3A0hnPYJ6dnyjrHEECm9tCQzJABrKVR5gHc7rlCp2MmAN1gqAARmgzDAoFtUEhXGJUAB9ahgKAAmBeVBcNgdBacIQgDhwEBQDioJEokqY8bAGRAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false).

For all of these reasons I am going to reject this proposal -- thanks for filing!

**Comment by dmurvihill** [comment]
Thanks, that's helpful!
